feat: even more precision for bash steps in github-env

ubiratansoares opened the pull request:
A follow-up for #197 

Here we leverage [tree-sitter-bash](https://crates.io/crates/tree-sitter-bash) rather than regexes to match usages of `$GITHUB_ENV` within a Bash script body. 

I took a simple approach, getting the full expansion for a `file_redirect` or `pipeline` and matching inner contents. Perhaps I'm missing some corner cases, but at least we have a starting point to iterate over 🙂

I also added some additional test cases (comments) and prepared the implementation to match against other types of shells in upcoming PRs.

woodruffw commented on src/audit/github_env.rs:
```diff
@@ -2,28 +2,66 @@ use super::{audit_meta, WorkflowAudit};
-    fn uses_github_environment(run_step_body: &str) -> bool {
-        GITHUB_ENV_WRITE_SHELL.is_match(run_step_body)
+    fn evaluate_github_environment_within_bash_script(script_body: &str) -> anyhow::Result<bool> {
+        let bash = tree_sitter_bash::LANGUAGE;
+        let mut parser = Parser::new();
```
Nit: creating the parser with each evaluation is probably slower than necessary; we could probably reuse `parser` by making it a property of `GitHubEnv` that gets loaded when the audit itself is loaded with `GitHubEnv::new` 🙂

ubiratansoares replied:
Done [here](https://github.com/woodruffw/zizmor/pull/208/commits/bc0e914233b4ac5209822e04f25faee2df6b573f), along with other changes!

woodruffw commented on src/audit/github_env.rs:
```diff
@@ -47,8 +85,9 @@ impl WorkflowAudit for GitHubEnv {
-        if let StepBody::Run { run, .. } = &step.deref().body {
-            if Self::uses_github_environment(run) {
+        if let StepBody::Run { run, shell, .. } = &step.deref().body {
+            let interpreter = shell.clone().unwrap_or("bash".into());
```
This is unfortunately not guaranteed to be the right default, since Windows runners default to `pwsh`. I think what we need to do is look at the surrounding `NormalJob::runs_on` and maybe add some kind of `RunsOn::default_shell()` helper. I'll look into that a bit.

(Nonblocker, since this is still improving on the precision of the current approach 🙂)

woodruffw replied:
Actually, it's a bit more annoying than even that: we need to determine the step's shell via the following precedence:

1. [defaults.run.shell](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#defaultsrunshell)
2. [jobs.<job_id>.defaults.run.shell](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#jobsjob_iddefaultsrunshell)
3. [jobs.<job_id>.steps[*].shell](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsshell)
4. If none of the above are set, use the default shell implied by the runner from `runs-on` (which might be indeterminate in the self-hosted runner case)

ubiratansoares replied:
@woodruffw I totally missed `pwsh` on Windows as the default, I never used Windows runners and I though that `bash` was the default (via WLS). Another day, another learning I guess 🙂

woodruffw replied:
I forget about too 🙂 -- the changes in #213 should enable this, but I'll rebase off of your work so that it's not a blocker here.

woodruffw commented on src/audit/github_env.rs:
I think we can fold these two test functions into one, with `(case, expected: bool)` as the parametrization 🙂

ubiratansoares commented:
@woodruffw I lack the energy to work on anything else today, I'll resume the work on this PR tomorrow, addressing all your suggestions, and perhaps making the most of #213 🙂

Meanwhile, thanks again for the great review (as usual), it's a pleasure contributing to and learning from this project ❤️

woodruffw commented:
Thank you too! And there's no rush or urgency at all, it's all supposed to be open source fun 🙂

woodruffw reviewed (approved):
LGTM, thanks a ton @ubiratansoares!
